### PR TITLE
🏗 ✨ ❄️ Update cireporter to run on CircleCI and generate a junit report

### PR DESCRIPTION
**Pull request by rileyajones**
Continuing on from #33682 to collect test results from e2e tests in order to show metrics in CircleCI.
CircleCI can only interpret test results in junit xml format.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -279,6 +279,7 @@ jobs:
       - run:
           name: 'End-to-End Tests'
           command: node build-system/pr-check/e2e-tests.js
+      - save_artifacts
```
Comment (rsimha) [inline .circleci/config.yml:282]: I'm guessing this should be `store_test_output`?

File: build-system/tasks/e2e/index.js [hunk 2]
```diff
@@ -80,20 +80,28 @@ function createMocha_() {
   let reporter;
   if (argv.testnames || argv.watch) {
     reporter = '';
-  } else if (argv.report) {
+  } else if (argv.report || isCircleciBuild()) {
```
Comment (rsimha) [inline build-system/tasks/e2e/index.js:83]: Will this work for CI builds where `--report` is not being passed in?

File: build-system/tasks/e2e/mocha-ci-reporter.js [hunk 3]
```diff
@@ -28,6 +29,7 @@ function ciReporter(runner, options) {
   Base.call(this, runner, options);
   this._mochaDotsReporter = new MochaDotsReporter(runner, options);
   this._jsonReporter = new JsonReporter(runner, options);
+  this._mochaJunitReporter = new MochaJUnitReporter(runner, options);
```
Comment (rsimha) [inline build-system/tasks/e2e/mocha-ci-reporter.js:32]: Clarifying question: Is this the change that was requested by https://github.com/ampproject/amphtml/pull/33682#discussion_r611716106?

**Review by rsimha — COMMENTED** [review]
A few comments below. Will defer to Esther to review / approve the reporter changes.

File: build-system/tasks/e2e/mocha-ci-reporter.js (hunk 3 above)
Comment (rileyajones) [reply build-system/tasks/e2e/mocha-ci-reporter.js:32]: Yes

File: .circleci/config.yml (hunk 1 above)
Comment (rileyajones) [reply .circleci/config.yml:282]: Indeed it should, rebase got me

File: build-system/tasks/e2e/index.js (hunk 2 above)
Comment (rileyajones) [reply build-system/tasks/e2e/index.js:83]: That is my intention, but I can see why that may be unintuitive. An alternative would be to set the variable in CircleCI (either in the config.yml file for each task or on the dashboard). Do you have a preference?

Comment (rsimha) [reply build-system/tasks/e2e/index.js:83]: On looking at the code, it looks like `argv.report` only generates the report file and doesn't affect the run in any other way, so I'm okay with leaving this as-is.

File: build-system/tasks/e2e/index.js [hunk 4]
```diff
@@ -33,7 +33,7 @@ const {buildRuntime, getFilesFromArgv} = require('../../common/utils');
 const {cyan} = require('kleur/colors');
 const {execOrDie} = require('../../common/exec');
 const {HOST, PORT, startServer, stopServer} = require('../serve');
-const {isCiBuild} = require('../../common/ci');
+const {isCiBuild, isCircleciBuild} = require('../../common/ci');
```
Comment (estherkim) [inline build-system/tasks/e2e/index.js:36]: I think it should be ok to reuse `isCiBuild` even though it's less precise.

File: build-system/tasks/e2e/index.js [hunk 5]
```diff
@@ -80,20 +80,28 @@ function createMocha_() {
...
     reporter = '';
-  } else if (argv.report) {
+  } else if (argv.report || isCircleciBuild()) {
     reporter = ciReporter;
   } else {
     reporter = dotsReporter;
   }
 
-  return new Mocha({
+  const options = {
```
Comment (estherkim) [inline build-system/tasks/e2e/index.js:89]: Can this stay the same but with `reporterOptions: isCiBuild() ?? {...}`

Comment (rileyajones) [reply build-system/tasks/e2e/index.js:89]: That should be fine but `reporterOptions` will always exist this way.

**Review by estherkim — APPROVED** [review]

**Pull request merged**